apluslms_roman/utils/env.py

```diff
@@ -0,0 +1,275 @@
+
+
+def to_str(data):
+    if data is None or isinstance(data, str):
+        return data
+    if hasattr(data, 'get_data'):
+        data = data.get_data()
+    return to_json(data)
+
+
+def default_value(var, default):
+    return var if var is not None else default
+
+
+def alt_value(var, alternative):
+    return alternative if var is not None else ""
+
+
+def replace_once(var, pattern):
+    return replace(var, pattern, 1)
+
+
+def replace_all(var, pattern):
+    return replace(var, pattern, 0)
+
+
+def replace(var, pattern, count):
+    var = to_str(var)
+    try:
+        regex, replacement = parse_replacement_pattern(pattern)
+    except ValueError:
+        sep = '/' if count else '//'
+        raise EnvError(_(
+            "Wrong pattern replacement syntax. The format "
+            "should be ${{var{}pattern/replacement}}. ").format(sep))
+    return re.sub(regex, replacement, var, count=count)
+
+
+# separators currently recognized by the script
+SEPARATORS = OrderedDict((
+    (":-", {'func': default_value, 'allow_missing': True}),
+    (":+", {'func': alt_value, 'allow_missing': True}),
+    ("//", {'func': replace_all, 'allow_missing': False}),
+    ("/", {'func': replace_once, 'allow_missing': False}),
+))
+
+
+def parse_replacement_pattern(pattern):
+    matches = [rgx for rgx in replacement_rgxs if re.fullmatch(rgx, pattern)]
+    if not matches:
+        raise ValueError
+
+    # no extra '/'
+    if replacement_rgxs[0] in matches:
+        regex, replacement = pattern.split("/")
+    # both parts are quoted
+    elif replacement_rgxs[1] in matches:
+        # the 'quoted' regex returns a tuple as a match object because
+        # there are two options ('' or  "")
+        regex, replacement = re.findall(quoted_rgx, pattern)
+        regex = regex[0] or regex[1]
+        replacement = replacement[0] or replacement[1]
+    # first part is quoted
+    elif replacement_rgxs[2] in matches:
+        regex = re.search(quoted_rgx, pattern).group(0)
+        replacement = pattern.rpartition('/')[2]
+    # second part is quoted
+    else:
+        replacement = re.search(quoted_rgx, pattern).group(0)
+        regex = pattern.partition('/')[0]
+
+    regex = regex.strip('\'"')
+    replacement = replacement.strip('\'"')
+
+    return regex, replacement
+
+
+def update(env, key, value):
+    if isinstance(value, str):
+        value = get_val(env, key, value)
+    elif isinstance(value, (list, Sequence)):
+        for i, item in enumerate(value):
+            value[i] = get_val(env, key, item)
+    elif isinstance(value, (dict, MutableMapping)):
+        for key_, item in value.items():
+            value[key_] = get_val(env, key, item)
+
+    env[key] = value
+
+
+def get_val(env, key, value):
+    references_self = _("Variable {} references itself")
+    not_defined = _("{} hasn't been defined")
+
+    if isinstance(value, int):
+        return value
+    value = to_str(value)
+    while True:
+        var = re.search(sub_rgx, value)
```

Comment:
Yep. This is the place that would be nicer with the assignment expressions https://www.python.org/dev/peps/pep-0572/, sadly we can't use it yet :(